### ISSUE TITLE
Fast booting on 3DS systems using .firm/BOOT file for TwiLight Menu++

### ISSUE THREAD

**Issue by SebsterDev** [DS-Homebrew/TWiLightMenu#2495]
**Is your feature request related to a problem? Please describe.**
I use my 3DS a lot for playing all kinds of games I've missed out on for the DS systems, but I find the normal way of booting DS games to be slow and tedious... and I own only a N2dsXL system. Can't imagine how it might be on "old" models...

**Describe the solution you'd like**
I got this idea of making a .firm program for TwiLight when i found out about "FastBoot3DS" and "open_agb_firm". Through open_agb_firm I have become able to launch my GBA games _even faster_ than my "_New_" system boots up to the main 3DS menu. I was wondering if this could make the mildly infuriating start-up times for DS games much faster, if theoretically, it would be possible in some way, to make a .firm program in the TwiLight Menu folder to boot the menu faster during the DS's startup? It could then boot directly into TwiLight through holding the power button and another button at the same time. One could also download FastBoot3DS for this, as that has made it possible for me to boot open_agb_firm through holding Power+L simultaneously

**Describe alternatives you've considered**
Now, I myself haven't programmed anything in the past for the 3DS systems before, except for configuring some installed emulators, so I'm not sure if this would be possible at all, but an alternative to the menu booting would be to code a new menu like open_agb_firm, where it would display a text-based list over all the downloaded .nds ROMS on the MicroSD and proceed to boot the selected rom in DS mode instantly, or to program something like GodMode9 that boots up while holding the power button and a different button (like L) simultaneously.

**Additional context**
I feel like the 3DS systems' DS mode might be capable - to an extend - of cooperating with .firm programs.
The reason I have this idea is that I've recently found sort of a 'hack' for making the process of shutting down a 3DS system in DS mode much faster after a session of playing DS games... I did this by booting open_agb_firm while pressing the power button. Then shutting my system down on the open_agb_firm menu. Shutting down my 3DS after a session in DS mode now takes about a mere second. This could be a stretch, but this _could_ prove TwilightMenu able to interact with boot menus in some way.

**Comment by RocketRobz** [comment]
Have you tried using the Luma autoboot feature in TWLMenu++ Settings?

**Comment by SebsterDev** [comment]
> Have you tried using the Luma autoboot feature in TWLMenu++ Settings?

Is there any way to launch that in pixel-perfect mode? 
_But no... I just now found about what it did... i feel like it isn't very well described literally anywhere...?_
And what about my ideas for alternatives? would it be benefitial to have a text menu like open_agb_firm that boots faster to display the list of roms in the folder

**Comment by SebsterDev** [comment]
> Have you tried using the Luma autoboot feature in TWLMenu++ Settings?

Okay i've tried it for a few days now... now i have the problem that my 3DS boots up in DS mode when i'm not pressing on the D-pad at all...

**Comment by SebsterDev** [comment]
> Have you tried using the Luma autoboot feature in TWLMenu++ Settings?
@RocketRobz Sure, but i'd love to have a way to decide whenever i want to launch DSi mode and 3DS mode. I would appreciate if there was a .firm in the files for the menu, that can be interacted with through FastBoot3DS. I hope you take my advice, thanks.

**Comment by RocketRobz** [comment]
> @RocketRobz Sure, but i'd love to have a way to decide whenever i want to launch DSi mode and 3DS mode. I would appreciate if there was a .firm in the files for the menu, that can be interacted with through FastBoot3DS. I hope you take my advice, thanks.

This is something that'd have to be implemented within Luma3DS, as it handles starting both DS(i) and 3DS modes.

**Comment by SebsterDev** [comment]
> I modified Luma to ignore the autoboot setting in the configuration file and always autoboot into DSi mode. You can use that to set up a boot entry in fastboot3DS that always boots directly into DSi mode (TwilightMenu here) while your main boot entry boots into 3DS mode.
> 
> https://github.com/Ammako/Luma3DS/releases
> 
> You can use that until/if Luma ever implements an autoboot mode selector on their own.

Yo sorry i almost never check my GitHub, that's actually goated, thanks a lot that workaround sounds hella convenient, i'll look into that!

**Comment by SebsterDev** [comment]
> I modified Luma to ignore the autoboot setting in the configuration file and always autoboot into DSi mode. You can use that to set up a boot entry in fastboot3DS that always boots directly into DSi mode (TwilightMenu here) while your main boot entry boots into 3DS mode.
> 
> https://github.com/Ammako/Luma3DS/releases
> 
> You can use that until/if Luma ever implements an autoboot mode selector on their own.

Where do i put your .firm on my SD?

**Comment by SebsterDev** [comment]
> I imagine you would get the same problem if you autobooted into DSi mode using regular Luma, because the only changes in mine is that the autoboot settings are ignored and always autoboots into DSi mode.

Okay i might have missed some steps then. How did you mark the autoboot settings as ignored? Because i just marked the line in quotation but i don't know if that's the correct way to do it. Also do you know why would it make the crash error when i try to boot in DSi mode? Wdym by "regular luma", should i replace luma with your file in some way? I think i need to know what the differences are between my luma and your luma as of now..

**Comment by SebsterDev** [comment]
> It's in the code. You don't have to modify anything on your end.
> 
> > Configure DSi homebrew autoboot as normal, leave autoboot to Off in Luma settings, then configure your bootloader to run the official Luma boot.firm when you want to boot into 3DS mode, or dsi.firm when you want to boot into DSi mode.

Where do i put my dsi.firm? Just anywhere in my SD?

**Comment by SebsterDev** [comment]
> You put it anywhere you want. Then in fastboot3DS you just have to make a boot entry for it. There is a file selector to select your .firm for that boot entry, and you set a different key combo for that boot entry.
> 
> So for example you could have boot.firm (official luma) as the default boot entry that loads when not pressing anything, and a second boot entry that activates when you hold R on power on that runs dsi.firm instead.
> 
> Also, I thought the ds-homebrew wiki had the instructions for setting up DSi autoboot with Luma, but apparently it doesn't (or I can't find it.) You have to run Twilight Menu on your 3DS through regular Luma first, and in the Twilight Menu settings, there should be an option somewhere to set up Luma autoboot. This edits the Luma config to put Twilight Menu's title ID as the DSi autoboot title ID, so that it can properly autoboot into Twilight Menu.
> 
> This is just a quick hack, by the way. I would take the time to make a proper 3DS or DSi boot selector in Luma by holding a key combo, with a way to configure it directly in Luma's config menu, but I don't want to do all that work if they were already working on it themselves, and I haven't had a reply yet on whether or not they were interested in a PR. So I threw this together as a band-aid fix in the meantime.

Ohhh right i had the setting disabled, ever since i decided that i wanted to boot it through the regular 3DS menu anyways. It's very cool that you made this file, might not be the most polished or the real deal or anything, but it surely works how it should, and i think it's amazing! 
Only problem I have with it is the same thing i experienced with Luma's DSi-bootup mode. That's when i boot a different menu like Luma ingame from DSi-mode (which is something i do to turn off the console much faster lol) i need to boot my console twice the next time i want to boot in DSi-mode. Don't know if you've experienced this too...

Now i just gotta figure out why my TwPatch isn't working how it should, because with this out of the way, my screen filter not being applied is the only thing stopping me from playing DS games without tinkering or other start-up processes.

Thank you for making this! Wondering if i should close the thread here, as my current problem doesn't have anything to do with this.

**Comment by SebsterDev** [comment]
Right, makes sense. Hey, hope you don't mind me asking - but how come you made this, if you don't use DSi-mode?

**Comment by SebsterDev** [comment]
Dang, you're a legend!
